rust/flexbuffers/src/reader/mod.rs

```diff
@@ -0,0 +1,679 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::bitwidth::BitWidth;
+use crate::flexbuffer_type::FlexBufferType;
+use crate::Blob;
+use byteorder::{ByteOrder, LittleEndian};
+use std::convert::{From, TryFrom, TryInto};
+use std::ops::Rem;
+use std::str::FromStr;
+mod de;
+mod iter;
+mod map;
+mod vector;
+pub use iter::ReaderIterator;
+pub use map::{MapReader, MapReaderIndexer};
+pub use vector::VectorReader;
+
+/// All the possible errors when reading a flexbuffer.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
```

Comment:
Still don't agree that most of these should even exist. The C++ API is entirely error-less, which I believe is the correct design. But I'll let @rw decide.

Reply:
@aardappel Does the C++ FlexBuffers API have any UB? I wonder if the Rust API needs Error types to be safer.

Reply:
I think these should exist because rust people will want the safe option. We do have an API for failing default as per C++

Reply:
@rw UB, as in, if you call this method on the wrong type you'll get unexpected bits? no, quite the opposite.

@CasperN "rust people will want the safe option". This is not about safety vs unsafe. This is about API design. You are accessing fundamentally "dynamically typed" data, and no Rust API is ever going to make that "safe". It is not "unsafe" as in the Rust keyword either, since it is not untyped.

Put in this way: much like JSON, the data cannot be forced to be conforming to certain types, that is the whole point of its design (as an alternative to the strongly typed FlatBuffers). So trying to make the API that accesses it as strongly typed produces an API that will access data in a brittle way. Data should not crash code, code should be resilient to data changes, especially if the way that data is defined says "you can store whatever here, no need to conform to a schema". That's exactly what the C++ code does.

C++ users and Rust users both work in mostly strongly typed languages, and typically prefer to receive strongly typed data. But here, they are receiving data that *by design* can be anything. The API needs to be graceful when that anything happens. If you want to write code that says "this thing must be a float or die" then you should be using FlatBuffers instead :)

Reply:
Yea I don't mean unsafe as in undefined behavior. I think we should expose an API that does not discard information and instead provide it in the type. Sometimes people need to know when zero really means zero or if it means an error occurred. 

The [rust json library](https://docs.serde.rs/serde_json/value/enum.Value.html) returns options to indicate when type casting fails. That's an example of data that can be anything that is exposed with types that forces the programmer to call `unwrap_or_default` explicitly if want a zero-like value.

I think the current API has a reasonable tradeoff:

- `get_u32` does the strict annoying result type. 
- `as_u32` will do the casting and stuff as per the C++ behavior to return a u32 (or return 0).

This API shouldn't die either, unless there's a bug on my end

Reply:
Ok, I am not going to repeat myself, but I don't think you're understanding what I said above. If Rust users want brittle, untestable code that will fall over on user data changes, so be it. Merge at will.

Reply:
Ok, we agree to disagree. I have the C++ like API and the resulty API that's more similar to rust's serde_json